File: app/src/main/java/com/elishaazaria/sayboard/ime/ViewManager.java
package com.elishaazaria.sayboard.ime;

import android.annotation.SuppressLint;
import android.content.res.ColorStateList;
import android.content.res.Configuration;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.widget.TextViewCompat;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import com.elishaazaria.sayboard.R;
import com.elishaazaria.sayboard.ime.recognizers.RecognizerState;
import com.elishaazaria.sayboard.preferences.UIPreferences;

public class ViewManager implements Observer<RecognizerState> {
    private final IME ime;

    public static final int STATE_INITIAL = 0;
    public static final int STATE_LOADING = 1;
    public static final int STATE_READY = 2; // model loaded, ready to start
    public static final int STATE_LISTENING = 3;
    public static final int STATE_PAUSED = 4;
    public static final int STATE_ERROR = 5;

    private ConstraintLayout overlayView;
    private ImageButton micButton;
    private ImageButton backButton;
    private ImageButton backspaceButton;
    private ImageButton returnButton;
    private Button modelButton;
    private TextView resultView;

    public final MutableLiveData<Integer> stateLD = new MutableLiveData<>(STATE_INITIAL);
    public final MutableLiveData<Integer> errorMessageLD = new MutableLiveData<>(0);
    public final MutableLiveData<String> recognizerNameLD = new MutableLiveData<>("");

    private Listener listener;

    public ViewManager(IME ime) {
        this.ime = ime;

        stateLD.observe(ime, this::observeState);
        errorMessageLD.observe(ime, this::observeError);
        recognizerNameLD.observe(ime, this::observeRecognizerName);
    }

    public void init() {
        initializeVariables();

        reloadOrientation();

        setUpListeners();


        currentForeground = Integer.MAX_VALUE;
        currentBackground = Integer.MAX_VALUE;
        setUpTheme();
    }

    @SuppressLint("InflateParams")
    private void initializeVariables() {
        overlayView = (ConstraintLayout) ime.getLayoutInflater().inflate(R.layout.ime, null);
        resultView = overlayView.findViewById(R.id.result_text);
        micButton = overlayView.findViewById(R.id.mic_button);
        backButton = overlayView.findViewById(R.id.back_button);
        backspaceButton = overlayView.findViewById(R.id.backspace_button);
        modelButton = overlayView.findViewById(R.id.model_button);
        returnButton = overlayView.findViewById(R.id.return_button);

        resultView.setMovementMethod(new ScrollingMovementMethod());
    }

    private void setUpListeners() {
        micButton.setOnClickListener(v -> {
            if (listener != null) listener.micClick();
        });
        micButton.setOnLongClickListener(v -> listener != null && listener.micLongClick());
        backButton.setOnClickListener(v -> {
            if (listener != null) listener.backClicked();
        });
        backspaceButton.setOnClickListener(v -> {
            if (listener != null) listener.backspaceClicked();
        });
        backspaceButton.setOnTouchListener((v, event) -> {
            if (listener == null) {
                v.performClick();
                return false;
            }
            return listener.backspaceTouched(v, event);
        });
        returnButton.setOnClickListener(v -> {
            if (listener != null) listener.returnClicked();
        });
        modelButton.setOnClickListener(v -> {
            if (listener != null) listener.modelClicked();
        });
    }

    private int currentForeground = Integer.MAX_VALUE;
    private int currentBackground = Integer.MAX_VALUE;

    private void setUpTheme() {
        boolean dark = (ime.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES;

        int foreground = UIPreferences.getForegroundColor(dark, ime);
        int background = UIPreferences.getBackgroundColor(dark);

        if (currentForeground == foreground && currentBackground == background) return;

        currentForeground = foreground;
        currentBackground = background;

        overlayView.setBackgroundColor(background);

        ColorStateList foregroundTint = ColorStateList.valueOf(foreground);
        micButton.setImageTintList(foregroundTint);
        backButton.setImageTintList(foregroundTint);
        backspaceButton.setImageTintList(foregroundTint);
        returnButton.setImageTintList(foregroundTint);
        TextViewCompat.setCompoundDrawableTintList(modelButton, foregroundTint);
        modelButton.setTextColor(foreground);
        resultView.setTextColor(foreground);
    }

    private void reloadOrientation() {
        boolean landscape = ime.getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;

        Window window = ime.getMyWindow();
        if (window == null) return;

        int screenHeight = ime.getResources().getDisplayMetrics().heightPixels;

        float percent;
        if (landscape) {
            percent = UIPreferences.getScreenHeightLandscape();
        } else {
            percent = UIPreferences.getScreenHeightPortrait();
        }
        int height = (int) (percent * screenHeight);


        Log.d("ViewManager", "Screen height: " + screenHeight + ", height: " + height);

//        WindowManager.LayoutParams params = window.getAttributes();
//        params.height = height;
//        window.setAttributes(params);
//
        overlayView.setMinHeight(height);
        overlayView.setMaxHeight(height);
//        overlayView.setLayoutParams(new FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, height));
    }

    public void refresh() {
        setUpTheme();

        reloadOrientation();
    }

    private void observeState(int state) {
        boolean enabled;
        int text;
        int icon;
        switch (state) {
            case STATE_INITIAL:
            case STATE_LOADING:
                text = R.string.mic_info_preparing;
                icon = R.drawable.ic_settings_voice;
                enabled = false;
                break;
            case STATE_READY:
            case STATE_PAUSED:
                text = R.string.mic_info_ready;
                icon = R.drawable.ic_mic_none;
                enabled = true;
                break;
            case STATE_LISTENING:
                text = R.string.mic_info_recording;
                icon = R.drawable.ic_mic;
                enabled = true;
                break;
            case STATE_ERROR:
                text = R.string.mic_info_error;
                icon = R.drawable.ic_mic_off;
                enabled = false;
                break;
            default:
                return;
        }
        resultView.setText(text);
        micButton.setImageDrawable(AppCompatResources.getDrawable(ime, icon));
        micButton.setEnabled(enabled);
    }

    private void observeError(int messageId) {
        if (messageId == 0) return;
        resultView.setText(ime.getText(messageId));
    }

    public void observeRecognizerName(String name) {
        modelButton.setText(name);
    }

    public ConstraintLayout getRoot() {
        return overlayView;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public interface Listener {
        void micClick();

        boolean micLongClick();

        void backClicked();

        void backspaceClicked();

        boolean backspaceTouched(View v, MotionEvent event);

        void returnClicked();

        void modelClicked();
    }

    @Override
    public void onChanged(RecognizerState recognizerState) {
        switch (recognizerState) {
            case CLOSED:
            case NONE:
                stateLD.setValue(STATE_INITIAL);
                break;
            case LOADING:
                stateLD.setValue(STATE_LOADING);
                break;
            case READY:
                stateLD.setValue(STATE_READY);
                break;
            case IN_RAM:
                stateLD.setValue(STATE_PAUSED);
                break;
            case ERROR:
                stateLD.setValue(STATE_ERROR);
                break;
        }
    }

    private int convertDpToPixel(float dp) {
        return (int) (dp * (ime.getResources().getDisplayMetrics().densityDpi / 160f));
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/ime/ViewManager.kt
package com.elishaazaria.sayboard.ime

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.util.Log
import android.view.inputmethod.EditorInfo
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowColumn
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Icon
import androidx.compose.material.IconButton
import androidx.compose.material.LocalContentAlpha
import androidx.compose.material.LocalContentColor
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.darkColors
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowBack
import androidx.compose.material.icons.filled.ArrowRightAlt
import androidx.compose.material.icons.filled.Backspace
import androidx.compose.material.icons.filled.KeyboardReturn
import androidx.compose.material.icons.filled.Language
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.MicNone
import androidx.compose.material.icons.filled.MicOff
import androidx.compose.material.icons.filled.NavigateBefore
import androidx.compose.material.icons.filled.NavigateNext
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Send
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.SettingsVoice
import androidx.compose.material.lightColors
import androidx.compose.material.minimumInteractiveComponentSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.PointerInputChange
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.AbstractComposeView
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.elishaazaria.sayboard.AppPrefs
import com.elishaazaria.sayboard.Constants
import com.elishaazaria.sayboard.R
import com.elishaazaria.sayboard.recognition.recognizers.RecognizerState
import com.elishaazaria.sayboard.sayboardPreferenceModel
import com.elishaazaria.sayboard.theme.Shapes
import com.elishaazaria.sayboard.ui.utils.MyIconButton
import com.elishaazaria.sayboard.ui.utils.MyTextButton
import dev.patrickgold.jetpref.datastore.model.observeAsState
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@SuppressLint("ViewConstructor")
class ViewManager(private val ime: Context) : AbstractComposeView(ime),
    Observer<RecognizerState> {
    private val prefs by sayboardPreferenceModel()
    val stateLD = MutableLiveData(STATE_INITIAL)
    val errorMessageLD = MutableLiveData(R.string.mic_info_error)
    private var listener: Listener? = null
    val recognizerNameLD = MutableLiveData("")
    val enterActionLD = MutableLiveData(EditorInfo.IME_ACTION_UNSPECIFIED)

    init {
        layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
    }

    @OptIn(ExperimentalLayoutApi::class)
    @Composable
    override fun Content() {
        val stateS = stateLD.observeAsState()
        val errorMessageS = errorMessageLD.observeAsState(R.string.mic_info_error)
        val recognizerNameS = recognizerNameLD.observeAsState(initial = "")
        val height =
            (LocalConfiguration.current.screenHeightDp * when (LocalConfiguration.current.orientation) {
                Configuration.ORIENTATION_LANDSCAPE -> prefs.keyboardHeightLandscape.get()
                else -> prefs.keyboardHeightPortrait.get()
            }).toInt().dp
        IMETheme(prefs) {
            CompositionLocalProvider(
                LocalContentColor provides MaterialTheme.colors.primary
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(height)
                        .background(MaterialTheme.colors.background)
                ) {
                    Column {
                        Row(verticalAlignment = Alignment.Top) {
                            IconButton(onClick = { listener?.backClicked() }) {
                                Icon(
                                    imageVector = Icons.Default.ArrowBack,
                                    contentDescription = null
                                )
                            }
                            val topKeys by prefs.keyboardKeysTop.observeAsState()
                            FlowRow(
                                modifier = Modifier.weight(1f),
                                horizontalArrangement = Arrangement.Center
                            ) {
                                for (key in topKeys) {
                                    MyTextButton(onClick = { listener?.buttonClicked(key.text) }) {
                                        Text(text = key.label)
                                    }
                                }
                            }
                            Box(
                                modifier = Modifier
                                    .pointerInput(Unit) {
                                        detectTapGestures(onPress = {
                                            var down = true;
                                            coroutineScope {
                                                val repeatJob = launch {
                                                    delay(Constants.BackspaceRepeatStartDelay)
                                                    while (down) {
                                                        listener?.backspaceClicked()
                                                        delay(Constants.BackspaceRepeatDelay)
                                                    }
                                                }
                                                launch {
                                                    val released = tryAwaitRelease();
                                                    down = false;
                                                    Log.d("ViewManager", "$released")
                                                    repeatJob.cancel()
                                                }
                                            }
                                        }, onTap = {
                                            listener?.backspaceClicked()
                                        })
                                    }
                                    .pointerInput(Unit) {
                                        detectHorizontalDragGestures(onDragStart = {
                                            listener?.backspaceTouchStart(it)
                                        }, onDragCancel = {
                                            listener?.backspaceTouchEnd()
                                        }, onDragEnd = {
                                            listener?.backspaceTouchEnd()
                                        }, onHorizontalDrag = { change, amount ->
                                            listener?.backspaceTouched(change, amount)
                                        })
                                    }
                                    .minimumInteractiveComponentSize()

                                    .padding(vertical = 8.dp, horizontal = 16.dp),
                                contentAlignment = Alignment.Center
                            ) {
                                val contentAlpha = LocalContentAlpha.current
                                CompositionLocalProvider(LocalContentAlpha provides contentAlpha) {
                                    Icon(
                                        imageVector = Icons.Default.Backspace,
                                        contentDescription = null
                                    )
                                }
                            }
                        }
                        Row(modifier = Modifier.weight(1f)) {
                            val leftKeys by prefs.keyboardKeysLeft.observeAsState()
                            FlowColumn() {
                                for (key in leftKeys) {
                                    MyTextButton(onClick = { listener?.buttonClicked(key.text) }) {
                                        Text(text = key.label)
                                    }
                                }
                            }
                            Column(
                                modifier = Modifier
                                    .weight(1f)
                                    .fillMaxSize()
                            ) {
                                MyIconButton(onClick = {
                                    listener?.micClick()
                                }, onLongClick = {
                                    listener?.micLongClick()
                                },
                                    modifier = Modifier
                                        .weight(1f)
                                        .fillMaxSize()
                                ) {
                                    Icon(
                                        imageVector = when (stateS.value) {
                                            STATE_INITIAL, STATE_LOADING -> Icons.Default.SettingsVoice
                                            STATE_READY, STATE_PAUSED -> Icons.Default.MicNone
                                            STATE_LISTENING -> Icons.Default.Mic
                                            else -> Icons.Default.MicOff
                                        }, contentDescription = null,
                                        modifier = Modifier.fillMaxSize()
                                    )
                                }
                                Text(
                                    text = when (stateS.value) {
                                        STATE_INITIAL, STATE_LOADING -> stringResource(id = R.string.mic_info_preparing)
                                        STATE_READY, STATE_PAUSED -> stringResource(id = R.string.mic_info_ready)
                                        STATE_LISTENING -> stringResource(id = R.string.mic_info_recording)
                                        else -> stringResource(id = errorMessageS.value)
                                    }, fontSize = MaterialTheme.typography.h6.fontSize,
                                    modifier = Modifier
                                        .align(Alignment.CenterHorizontally)
                                        .padding(5.dp)
                                )
                            }
                            val rightKeys by prefs.keyboardKeysRight.observeAsState()
                            FlowColumn {
                                for (key in rightKeys) {
                                    MyTextButton(onClick = { listener?.buttonClicked(key.text) }) {
                                        Text(text = key.label)
                                    }
                                }
                            }
                        }
                        Row {
                            IconButton(onClick = { listener?.settingsClicked() }) {
                                Icon(
                                    imageVector = Icons.Default.Settings,
                                    contentDescription = null
                                )
                            }
                            IconButton(
                                onClick = { listener?.modelClicked() }
                            ) {
                                Row {
                                    Icon(
                                        imageVector = Icons.Default.Language,
                                        contentDescription = null
                                    )
                                    Text(text = recognizerNameS.value)
                                }
                            }
                            Spacer(modifier = Modifier.weight(1f))
                            IconButton(onClick = { listener?.returnClicked() }) {
                                val enterAction by enterActionLD.observeAsState()
                                Icon(
                                    imageVector = when (enterAction) {
                                        EditorInfo.IME_ACTION_GO -> Icons.Default.ArrowRightAlt
                                        EditorInfo.IME_ACTION_SEARCH -> Icons.Default.Search
                                        EditorInfo.IME_ACTION_SEND -> Icons.Default.Send
                                        EditorInfo.IME_ACTION_NEXT -> Icons.Default.NavigateNext
                                        EditorInfo.IME_ACTION_PREVIOUS -> Icons.Default.NavigateBefore
                                        else -> Icons.Default.KeyboardReturn
                                    },
                                    contentDescription = null,
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    override fun onChanged(value: RecognizerState) {
        when (value) {
            RecognizerState.CLOSED, RecognizerState.NONE -> stateLD.setValue(STATE_INITIAL)

            RecognizerState.LOADING -> stateLD.setValue(STATE_LOADING)
            RecognizerState.READY -> stateLD.setValue(STATE_READY)
            RecognizerState.IN_RAM -> stateLD.setValue(STATE_PAUSED)
            RecognizerState.ERROR -> stateLD.setValue(STATE_ERROR)
        }
    }

    fun setListener(listener: Listener) {
        this.listener = listener
    }

    interface Listener {
        fun micClick()
        fun micLongClick(): Boolean
        fun backClicked()
        fun backspaceClicked()
        fun backspaceTouchStart(offset: Offset)
        fun backspaceTouched(change: PointerInputChange, dragAmount: Float)
        fun backspaceTouchEnd()
        fun returnClicked()
        fun modelClicked()
        fun settingsClicked()
        fun buttonClicked(text: String)
    }

    companion object {
        const val STATE_INITIAL = 0
        const val STATE_LOADING = 1
        const val STATE_READY = 2 // model loaded, ready to start
        const val STATE_LISTENING = 3
        const val STATE_PAUSED = 4
        const val STATE_ERROR = 5

//        const val BUTTON_TOP_LEFT = 1
//        const val BUTTON_MIDDLE_LEFT = 2
//        const val BUTTON_BOTTOM_LEFT = 3
//        const val BUTTON_TOP_RIGHT = 11
//        const val BUTTON_MIDDLE_RIGHT = 12
//        const val BUTTON_BOTTOM_RIGHT = 13

    }
}

@Composable
fun IMETheme(
    prefs: AppPrefs,
    content: @Composable () -> Unit
) {
    val colors = if (isSystemInDarkTheme()) {
        darkColors(
            background = Color(prefs.uiNightBackground.get()),
            primary = if (prefs.uiNightForegroundMaterialYou.get()) {
                colorResource(id = R.color.materialYouForeground)
            } else {
                Color(prefs.uiNightForeground.get())
            },
        )
    } else {
        lightColors(
            background = Color(prefs.uiDayBackground.get()),
            primary = if (prefs.uiDayForegroundMaterialYou.get()) {
                colorResource(id = R.color.materialYouForeground)
            } else {
                Color(prefs.uiDayForeground.get())
            },
        )
    }

    MaterialTheme(
        colors = colors,
        shapes = Shapes,
        content = content,
    )
}